### PR TITLE
Issue/62

### LINKED ISSUE #62: Possible to support OPTIONS requests for CORS?
**Is your feature request related to a problem? Please describe.**

Just tried this project out for the first time and it's great. Docs were clear and easy to get started.

One issue I ran into is that it appears OPTIONS requests currently are rejected by the JS module: https://github.com/nginxinc/nginx-s3-gateway/blob/master/common/etc/nginx/include/s3gateway.js#L410-L415

It would be great to be able to enable CORS settings since that would make it easier to use this proxy directly from the browser for client-side applications.

Is it possible to add support there, along with some means to configure CORS settings? 

**Describe the solution you'd like**

**Describe alternatives you've considered**

I tried adding CORS settings via an nginx conf file under `/etc/nginx/conf.d/` but I think the JS module is taking precedence and rejecting these requests before my settings take effect.

That said, it's definitely possible I am configuring something wrong...I am by no means an expert on either NGINX or CORS.

Here is the dockerfile I was using:

```
FROM ghcr.io/nginxinc/nginx-s3-gateway/nginx-oss-s3-gateway:latest

COPY nginx.conf /etc/nginx/conf.d/default.conf
```

And this is the config file i tried:

```
server {
    echo
    listen       80;
    server_name  localhost;

    #access_log  /var/log/nginx/host.access.log  main;

    location / {
        echo "hello"
        if ($request_method = 'OPTIONS') {
           echo "method is options"
           add_header 'Access-Control-Allow-Origin' '*';
           add_header 'Access-Control-Allow-Methods' 'GET, POST, OPTIONS';
           #
           # Custom headers and headers various browsers *should* be OK with but aren't
           #
           add_header 'Access-Control-Allow-Headers' 'DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range';
           #
           # Tell client that this pre-flight info is valid for 20 days
           #
           add_header 'Access-Control-Max-Age' 1728000;
           add_header 'Content-Type' 'text/plain; charset=utf-8';
           add_header 'Content-Length' 0;
           return 204;
        }
        if ($request_method = 'POST') {
            add_header 'Access-Control-Allow-Origin' '*' always;
            add_header 'Access-Control-Allow-Methods' 'GET, POST, OPTIONS' always;
            add_header 'Access-Control-Allow-Headers' 'DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range' always;
            add_header 'Access-Control-Expose-Headers' 'Content-Length,Content-Range' always;
        }
        if ($request_method = 'GET') {
            add_header 'Access-Control-Allow-Origin' '*' always;
            add_header 'Access-Control-Allow-Methods' 'GET, POST, OPTIONS' always;
            add_header 'Access-Control-Allow-Headers' 'DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range' always;
            add_header 'Access-Control-Expose-Headers' 'Content-Length,Content-Range' always;
        }

        # root   /usr/share/nginx/html;
        # index  index.html index.htm;

    }

    error_page   500 502 503 504  /50x.html;
    location = /50x.html {
        root   /usr/share/nginx/html;
    }
}
```

Apologies in advance if those are way off base, but maybe someone can steer me in a better direction if this is not the right way to configure these settings.

**Additional context**

Example:

```
nginx-proxy ➸ curl -I -X OPTIONS http://localhost/example.txt
HTTP/1.1 405 Not Allowed
Server: nginx
Date: Thu, 13 Oct 2022 17:33:32 GMT
Content-Type: text/html
Content-Length: 150
Connection: keep-alive
Allow: GET, HEAD
```

### PR DESCRIPTION
**Pull request by dekobon**
Fixes #62

Add support for CORS response headers.

### CODE REVIEW

**Review by 4141done — COMMENTED** [review]
I did see an issue that would lead to "if inside location block" in the `cors.conf` I think we should discuss before merging.

File: common/docker-entrypoint.sh
Comment (dekobon) [reply common/docker-entrypoint.sh:51]: Yeah, it seems unnecessary because this approach is stemming from my security paranoia. I don't want to provide any paths to allow for proxying requests to the underlying S3 bucket beyond the narrow operations defined. As such, I'm going out of my way here to lock down the HTTP verbs to only GET/HEAD or in the case of CORS GET/HEAD/OPTIONS.

File: common/etc/nginx/templates/gateway/cors.conf.template
Comment (dekobon) [reply common/etc/nginx/templates/gateway/cors.conf.template:17]: For the `OPTIONS` if, it seems to fall under the category of "good ifs":
> There are cases where you simply cannot avoid using an if, for example, if you need to test a variable which has no equivalent directive.

As for the other ifs, I believe they are ok due to there being no ifs after them. However, I have no idea how you would be able to do the same operation with map. Ideally, I wanted to conditionally template in the cors.conf include, but that seemed a bit too clunky. I guess I could conditional set an env var with the whole include statement in `docker-entrypoint.sh`.

File: common/etc/nginx/templates/gateway/cors.conf.template
Comment (4141done) [inline common/etc/nginx/templates/gateway/cors.conf.template:30]: Curious why `Access-Control-Expose-Headers` is not sent in the `OPTIONS` response?

File: common/etc/nginx/templates/gateway/cors.conf.template
Comment (4141done) [reply common/etc/nginx/templates/gateway/cors.conf.template:17]: Yeah this certainly is the best for understandibility.  I wouldn't add more moving parts if it's testing ok.   I'm going to play around with making it work with map because I'm curious now but let's not let it hold up the PR.

**Review by 4141done — APPROVED** [review]
👍 🐳 👍 
This one looks good to me - I have one outstanding question around a header included for all methods except `OPTIONS` but if that's not a big deal let's merge it 🐎

**Comment by dekobon** [comment]
@worace I'll be waiting a week for your review/test and then we will merge

**Pull request merged**
